scheme/arm-cca/realm.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2024 Contributors to the Veraison project.
+// SPDX-License-Identifier: Apache-2.0
+
+package arm_cca
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/veraison/ccatoken"
+	"github.com/veraison/ear"
+	"github.com/veraison/services/handler"
+	"github.com/veraison/services/log"
+	"github.com/veraison/services/scheme/common/arm"
+	realm "github.com/veraison/services/scheme/common/cca/realm"
+)
+
+type CCARealmAttester struct {
+}
+
+func (CCARealmAttester) PerformAppraisal(
+	appraisal *ear.Appraisal,
+	claimsMap map[string]interface{},
+	endorsements []handler.Endorsement) error {
+
+	realmEndorsements := arm.FilterRefVal(endorsements, "realm.reference-value")
+	claims, err := realm.MapToRealmClaims(claimsMap)
+	if err != nil {
+		return fmt.Errorf("unable to get realm claims from Realm claims map: %w", err)
+	}
+
+	// If crypto verification (including chaining) completes correctly,
+	// we can safely assume the Realm instance to be trustworthy
+	appraisal.TrustVector.InstanceIdentity = ear.TrustworthyInstanceClaim
+	appraisal.TrustVector.Executables = ear.UnrecognizedRuntimeClaim
+
+	for _, endorsement := range realmEndorsements {
+		rimRpvMatch := matchRimRpv(claims, &endorsement)
+		if rimRpvMatch {
+			appraisal.TrustVector.Executables = ear.ApprovedBootClaim
+			// Match REM's
+			remMatch := matchREMs(claims, &endorsement)
+			if remMatch {
+				appraisal.TrustVector.Executables = ear.ApprovedRuntimeClaim
+			}
+			break
+		} else {
+			// TO DO “UNRECOGNIZED_BOOT" is missing
+			appraisal.TrustVector.Executables = ear.ContraindicatedRuntimeClaim
+		}
+	}
+	appraisal.UpdateStatusFromTrustVector()
+	appraisal.VeraisonAnnotatedEvidence = &claimsMap
+
+	return nil
+}
+
+func matchRimRpv(claims ccatoken.IClaims, endorsement *handler.Endorsement) bool {
```

Comment:
Any specific reason to merge RIM and PV matching in the same function?

Reply:
Yes, if the Rim matches but RPV (if provided in Reference VALUE) does not match, then we could not identify the correct Realm instance correctly, hence they are matched together!

Reply:
Hmm, but that's not a necessity.  You could do:

```golang
func matchRIM(actual []byte, endorsement *handler.Endorsement) error {
    expected, err := realm.GetRim(endorsement.Attributes)
    if err != nil {
        return err
    }

    return bytes.Equal(actual, expected)
}

func matchRPV(actual []byte, endorsement *handler.Endorsement) error {
    expected, err := realm.GetRpv(endorsement.Attributes)
    if err != nil {
        if errors.Is(err, ErrNotFound) {
            return nil
        }
        return err
    }

    return bytes.Equal(actual, expected)
}
```
and then in the calling site:
```golang
    if matchRPV(rpv, endorsement) != nil && matchREM(rem, endorsement) != nil {
        match = true
    }
```

Reply:
True that, but at the end of the day they both are `intended` to identify `uniqueness of a Realm` , so logically compared together!

Reply:
Also RPV is Optional so in absence of RPV  `matchRPV` may fail, but that does not mean anything, so the comparison has to be in correct context, hence best to be done together!

Reply:
> True that, but at the end of the day they both are intended to identify uniqueness of a Realm , so logically compared together!

I don't see it as a logical necessity.  Also, my point is it makes the function body chunkier and therefore more difficult to grasp.

> Also RPV is Optional so in absence of RPV  matchRPV may fail, but that does not mean anything, so the comparison has to be in correct context, hence best to be done together!

Sure, but if you look at the (pseudo)code I proposed above, this bit:
```golang
        if errors.Is(err, ErrNotFound) {
            return nil
        }
```
deals with optionality.